File: llm_generated_code/ChatGPT/Python/Level1/Q33.py
def max_profit_with_two_transactions(prices):
    if not prices:
        return 0

    n = len(prices)
    left_profits = [0] * n
    right_profits = [0] * n

    min_price = prices[0]
    for i in range(1, n):
        min_price = min(min_price, prices[i])
        left_profits[i] = max(left_profits[i - 1], prices[i] - min_price)

    max_price = prices[-1]
    for i in range(n - 2, -1, -1):
        max_price = max(max_price, prices[i])
        right_profits[i] = max(right_profits[i + 1], max_price - prices[i])

    max_profit = 0
    for i in range(n):
        max_profit = max(max_profit, left_profits[i] + right_profits[i])

    return max_profit

prices = list(map(int, input().split()))
print(max_profit_with_two_transactions(prices))

File: llm_generated_code/ChatGPT/CPP/Level2/Q33.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int maxProfit(vector<int>& prices) {
    int n = prices.size();
    if (n <= 1) return 0;

    vector<int> profit1(n, 0);
    vector<int> profit2(n, 0);
    
    int minPrice = prices[0];
    for (int i = 1; i < n; ++i) {
        minPrice = min(minPrice, prices[i]);
        profit1[i] = max(profit1[i-1], prices[i] - minPrice);
    }
    
    int maxPrice = prices[n-1];
    for (int i = n - 2; i >= 0; --i) {
        maxPrice = max(maxPrice, prices[i]);
        profit2[i] = max(profit2[i+1], maxPrice - prices[i] + profit1[i]);
    }
    
    return profit2[0];
}

int main() {
    int t;
    cin >> t;
    
    while (t--) {
        int n;
        cin >> n;
        vector<int> prices(n);
        for (int i = 0; i < n; ++i) {
            cin >> prices[i];
        }
        cout << maxProfit(prices) << endl;
    }

    return 0;
}
